components/sidebar/sidebar_category_list/sidebar_category_list.tsx

```diff
@@ -2,32 +2,378 @@
+            }
+
+            this.scrollToPosition(scrollEnd);
+        } else if (bottom > scrollTop + scrollHeight) {
+            // Scroll down to the item
+            const margin = (scrollingToUnread || !this.state.showBottomUnread) ? scrollMargin : scrollMarginWithUnread;
+            const scrollEnd = (bottom - scrollHeight) + margin;
+
+            this.scrollToPosition(scrollEnd);
+        }
+    }
+
+    scrollToPosition = (scrollEnd: number) => {
+        // Stop the current animation before scrolling
+        this.scrollAnimation.setCurrentValue(this.scrollbar.current!.getScrollTop()).setAtRest();
+
+        this.scrollAnimation.setEndValue(scrollEnd);
+    }
+
+    updateUnreadIndicators = () => {
+        let showTopUnread = false;
+        let showBottomUnread = false;
+
+        // Consider partially obscured channels as above/below
+        const firstUnreadChannel = this.getFirstUnreadChannel();
+        const lastUnreadChannel = this.getLastUnreadChannel();
+
+        if (firstUnreadChannel) {
+            const firstUnreadElement = this.channelRefs.get(firstUnreadChannel);
+            const firstUnreadPosition = firstUnreadElement ? firstUnreadElement.offsetTop : null;
+
+            if (firstUnreadPosition && ((firstUnreadPosition + firstUnreadElement!.offsetHeight) - scrollMargin) < this.scrollbar.current!.getScrollTop()) {
+                showTopUnread = true;
+            }
+        }
+
+        if (lastUnreadChannel) {
+            const lastUnreadElement = this.channelRefs.get(lastUnreadChannel);
+            const lastUnreadPosition = lastUnreadElement ? lastUnreadElement.offsetTop : null;
+
+            if (lastUnreadPosition && (lastUnreadPosition + scrollMargin) > (this.scrollbar.current!.getScrollTop() + this.scrollbar.current!.getClientHeight())) {
+                showBottomUnread = true;
+            }
+        }
+        if (showTopUnread !== this.state.showTopUnread || showBottomUnread !== this.state.showBottomUnread) {
+            this.setState({
+                showTopUnread,
+                showBottomUnread,
+            });
+        }
+    }
+
+    getFirstUnreadChannel = () => {
+        return this.getFirstUnreadChannelFromChannelIdArray(this.getDisplayedChannels());
+    }
+
+    getLastUnreadChannel = () => {
+        return this.getFirstUnreadChannelFromChannelIdArray(this.getDisplayedChannels().reverse());
+    }
+
+    getDisplayedChannels = () => {
+        return this.props.categories.reduce((allChannelIds, section) => {
+            allChannelIds.push(...section.channel_ids);
+            return allChannelIds;
+        }, []);
+    }
+
+    navigateChannelShortcut = (e: KeyboardEvent) => {
+        if (e.altKey && !e.shiftKey && (Utils.isKeyPressed(e, Constants.KeyCodes.UP) || Utils.isKeyPressed(e, Constants.KeyCodes.DOWN))) {
+            e.preventDefault();
+
+            if (this.isSwitchingChannel) {
+                return;
+            }
+
+            this.isSwitchingChannel = true;
```

Comment:
Do we actually need `this.isSwitchingChannel` any more? The only way that it would matter is if we handle two key events at the same time which I think is impossible since JS is single-threaded and we don't pause to do anything asynchronous here

Reply:
I'll remove it, let's see what happens.